File: packages/wasmvm/wasmlib/src/wasmtypes/schname.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::convert::TryInto;

use crate::*;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub const SC_HNAME_LENGTH: usize = 4;

#[derive(PartialEq, Clone, Copy)]
pub struct ScHname(pub u32);

impl ScHname {
    pub fn new(name: &str]) -> ScHname {
        hname_from_bytes(&host::sandbox(FN_UTILS_HASH_NAME, &string_to_bytes(name)))
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        hname_to_bytes(*self)
    }

    pub fn to_string(&self) -> String {
        hname_to_string(*self)
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub fn hname_decode(dec: &mut WasmDecoder) -> ScHname {
    hname_from_bytes(&dec.fixed_bytes(SC_HNAME_LENGTH))
}

pub fn hname_encode(enc: &mut WasmEncoder, value: ScHname) {
    enc.fixed_bytes(&hname_to_bytes(value), SC_HNAME_LENGTH);
}

pub fn hname_from_bytes(buf: &[u8]) -> ScHname {
    if buf.len() == 0 {
        return ScHname(0);
    }
    if buf.len() != SC_HNAME_LENGTH {
        panic("invalid Hname length");
    }
    ScHname(u32::from_le_bytes(buf.try_into().expect("WTF?")))
}

pub fn hname_to_bytes(value: ScHname) -> Vec<u8> {
    value.0.to_le_bytes().to_vec()
}

pub fn hname_to_string(value: ScHname) -> String {
    let hexa = "0123456789abcdef".as_bytes();
    let mut res = [0u8; 8];
    let mut val = value.0;
    for n in 0..8 {
        res[7 - n] = hexa[val as usize & 0x0f];
        val >>= 4;
    }
    string_from_bytes(&res)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScImmutableHname {
    proxy: Proxy,
}

impl ScImmutableHname {
    pub fn new(proxy: Proxy) -> ScImmutableHname {
        ScImmutableHname { proxy }
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn to_string(&self) -> String {
        hname_to_string(self.value())
    }

    pub fn value(&self) -> ScHname {
        hname_from_bytes(&self.proxy.get())
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// value proxy for mutable ScHname in host container
pub struct ScMutableHname {
    proxy: Proxy,
}

impl ScMutableHname {
    pub fn new(proxy: Proxy) -> ScMutableHname {
        ScMutableHname { proxy }
    }

    pub fn delete(&self) {
        self.proxy.delete();
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn set_value(&self, value: ScHname) {
        self.proxy.set(&hname_to_bytes(value));
    }

    pub fn to_string(&self) -> String {
        hname_to_string(self.value())
    }

    pub fn value(&self) -> ScHname {
        hname_from_bytes(&self.proxy.get())
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/wasmtypes/schname.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {sandbox} from "../host";
import {FnUtilsHashName, panic} from "../sandbox";
import {uint32FromBytes, uint32ToBytes} from "./scuint32";
import {stringToBytes} from "./scstring";
import {bytesCompare} from "./scbytes";
import {WasmDecoder, WasmEncoder} from "./codec";
import {Proxy} from "./proxy";

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export const ScHnameLength = 4;

export class ScHname {
    id: Uint8Array;

    constructor(id: u32) {
        this.id = uint32ToBytes(id);
    }

    static fromName(name: string): ScHname {
        return hnameFromBytes(sandbox(FnUtilsHashName, stringToBytes(name)));
    }

    public equals(other: ScHname): bool {
        return bytesCompare(this.id, other.id) == 0;
    }

    // convert to byte array representation
    public toBytes(): Uint8Array {
        return hnameToBytes(this);
    }

    // human-readable string representation
    public toString(): string {
        return hnameToString(this)
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function hnameDecode(dec: WasmDecoder): ScHname {
    return hnameFromBytesUnchecked(dec.fixedBytes(ScHnameLength));
}

export function hnameEncode(enc: WasmEncoder, value: ScHname): void {
    enc.fixedBytes(value.toBytes(), ScHnameLength);
}

export function hnameFromBytes(buf: Uint8Array): ScHname {
    if (buf.length == 0) {
        return new ScHname(0);
    }
    if (buf.length != ScHnameLength) {
        panic("invalid Hname length");
    }
    return hnameFromBytesUnchecked(buf);
}

export function hnameToBytes(value: ScHname): Uint8Array {
    return value.id;
}

export function hnameFromString(value: string): ScHname {
    if (value.length > 8) {
        panic("invalid Hname string");
    }
    return new ScHname(parseInt(value, 16) as u32);
}

export function hnameToString(value: ScHname): string {
    const res = uint32FromBytes(value.id).toString(16);
    return "0000000".slice(0, 8 - res.length) + res;
}

function hnameFromBytesUnchecked(buf: Uint8Array): ScHname {
    let o = new ScHname(0);
    o.id = buf.slice(0);
    return o;
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableHname {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return hnameToString(this.value());
    }

    value(): ScHname {
        return hnameFromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableHname extends ScImmutableHname {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: ScHname): void {
        this.proxy.set(hnameToBytes(value));
    }
}
